### PR TITLE
[RNMobile] Gallery  - Native gallery component draft

### PR DESCRIPTION
**Pull request by mkevins**
## Description

This PR introduces a native `Gallery` component for the semi-cross-platform Gallery block. The purpose of this component is to render a list of images passed via the `images` prop.

### PR Hierarchy
This PR is part of the [PR hierarchy described here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1416#issuecomment-551057738). This PR can be tested with the aggregate changeset and integration of components within the related PRs by checking out the branch of the "top level" PR: https://github.com/WordPress/gutenberg/pull/18265 via the `gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1498.

### MediaPlaceholder

Changes to `MediaPlaceholder` from this PR: https://github.com/WordPress/gutenberg/pull/18262 have been merged for review here. The `MediaPlaceholder` component has been modified to append media for the native `Gallery` implementation.

On web, when new media is added to a gallery, the current selection of items in the gallery is available to the media selection interface, and can be presented as such to the user. The selector behaves more like a "gallery editor", in the sense that the selection is modified (users can add and / or remove items). On mobile, the interface for selecting media items is handled by the "parent apps", and does not "know" about what is currently selected as part of the gallery.

This PR adds the `addToGallery` prop to the `MediaPlaceholder` component, and conditionally appends media items to the current selection, passing the resulting collection back through `onSelect` when the `addToGallery` flag is set. When the `addToGallery` flag is not set, `onSelect` behavior is not modified.

**Note:** In this implementation, duplicates are removed from the collection via the `id` property, since this is used as a `key` in the list of elements that renders these items.
The MediaPlaceholder

The changeset here is used in the related "parent" PR: https://github.com/WordPress/gutenberg/pull/18111, which includes these changes integrated with related changes in other components necessary for the gallery block.

### Block-level caption

This component uses the mobile `Caption` component to add a block-level caption for gallery.

## To test
Test this component by checking out the branch of the "top level" PR: https://github.com/WordPress/gutenberg/pull/18265 via the `gutenberg-mobile` PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1498.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/block-library/src/gallery/gallery.native.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+	let displayedColumns;
+
+	// limit displayed columns based on viewport width
+	if ( isMobile ) {
+		displayedColumns = Math.min( columns, 2 );
+	} else if ( isNarrow ) {
+		displayedColumns = Math.min( columns, 4 );
+	} else {
+		displayedColumns = columns;
```
Comment (pinarol) [inline packages/block-library/src/gallery/gallery.native.js]: I gave it a try with iPad Pro 11 inch simulator, unfortunately our readable content width is still too narrow for 8 columns. Our readable content max width is 580.(see .../src/mobile/readable-content-view/style.native.scss) I don't know how isNarrow is calculated but that calculation should be based on readable content width and not screen width.


Screen Shot 2019-11-25 at 17 07 07

Comment (mkevins) [reply packages/block-library/src/gallery/gallery.native.js]: Thanks for pointing this out. I've pushed changes to set a `MAX_DISPLAYED_COLUMNS = 4` to limit the number of columns displayed for native. The displayed columns will still be limited to 2 for narrower screens.

Comment (pinarol) [reply packages/block-library/src/gallery/gallery.native.js]: I am fine with that fix currently 👍 In the long term we might need a more dynamic decision mechanism which can adjust itself accordingly due to the width of readable content. But, currently we have a constant limit there. So I don't see enough value to introduce such a dynamic mechanism today.

File: packages/block-library/src/gallery/gallery.native.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+	} = props;
+
+	const {
+		columns = defaultColumnsNumber( attributes ),
+		imageCrop,
+		images,
+	} = attributes;
+
+	// limit displayed columns based on viewport width
+	const displayedColumns = isMobile ?
```
Comment (pinarol) [inline packages/block-library/src/gallery/gallery.native.js]: Should we change `isMobile` as `isMobilePortrait` ? Because this is false for mobile landscape.

**Review by pinarol — APPROVED** [review]
I've tested this on both WPiOS and WPAndroid on device and working super good(except the multiple image selection on WPAndroid but that one is [handled separately](https://github.com/wordpress-mobile/WordPress-Android/pull/10857)) 🎉.  Also tested with dark mode on iOS and everything seems good. Great work! I am just entering a small comment.

One thing that is not great is VoiceOver support, it is saying lots of unnecessary things when I select a tile(I heard lots of 'horizontal scrollbar's). But I think we should tackle that on a separate PR, preferably before shipping the 1st iteration though.

File: packages/block-library/src/gallery/gallery.native.js (hunk 2 above)
Comment (mkevins) [reply packages/block-library/src/gallery/gallery.native.js]: Good point! `isMobile` is not the best name for this. I guess even using the name mobile in the the name at all can be misleading, since currently it's conditioned on viewport width being less than 600, while some larger tablets have a portrait viewport wider than that.

In an earlier iteration (when there were _3_ possibilities for displayed columns), I had passed another flag called `isNarrow`, which I think works semantically, but at that time may have been ambiguous (it wasn't clearly _more_ or _less_ narrow than `isMobile` for example). Since we're not using that anymore (and limiting to 4 columns on mobile in general because of the limited readable content width), maybe we could reuse that moniker? I.e. we could use the name `isNarrow: "< small"` (viewport width < 600), in place of `isMobile`? Wdyt?

**tl;dr:** Do you think `isNarrow` could work here?

Comment (pinarol) [reply packages/block-library/src/gallery/gallery.native.js]: `isNarrow` sounds OK to me.

Comment (pinarol) [reply packages/block-library/src/gallery/gallery.native.js]: It would really help to put a code comment somewhere about what range corresponds to isNarrow.

Comment (mkevins) [reply packages/block-library/src/gallery/gallery.native.js]: Good suggestion! I'll add a comment too. 😄
